### PR TITLE
Fix to weapon damage while shapeshifted

### PR DESCRIPTION
**Pull request by GuybrushGit**
Fixes https://github.com/LightsHope/server/issues/264

Both the weapon enchant damage and generic normal damage from SPELL_AURA_MOD_DAMAGE_DONE are being stored on UNIT_MOD_DAMAGE_MAINHAND -> TOTAL_VALUE. 
This makes it impossible to remove the enchant damage while on feral form since we want to only remove the weapon specific damage but not the generic.

I added a new unit mod UNIT_MOD_DAMAGE_PHYSICAL which just stores this generic damage, making it so UNIT_MOD_DAMAGE_MAINHAND will now only store weapon specific damage, which can then be disabled for ferals.

This also affects other classes when the weapon is broken / disarmed. Currently the enchant damage is not being removed for them as well.

### CODE REVIEW

**Comment by GuybrushGit** [comment]
Might of Cenarius will continue working since its generic damage not weapon specific, despite what the tooltip says.

**Comment by llopinator** [comment]
You're dismissing the exact wording of a tooltip and interpreting it in an entirely different way. If there's anything that the devs should give credence to, it's the wording of a tooltip, not someone's unsupported insistence that "oh, no that tooltip that says +4 weapon damage - it's REALLY generic damage because trust me."

If Might of Cenarius' +4 weapon damage does indeed work for druids (and we know it does), then that's unquestionable precedent for +9 weapon damage enchants working as well. Weightstones? Perhaps not, since patch 2.4.0 explicitly makes weightstones work while shapeshifted, and it's unclear whether the weightstones (a) worked in Vanilla then didn't early TBC, or (b) never worked until 2.4.0. 

Might of Cenarius is precedent for +9 weapon damage enchant working for druids. This is a no brainer.

**Comment by Nostalian** [comment]
You're dismissing the exact wording of a tooltip and interpreting it in an entirely different way. If there's anything that the devs should give credence to, it's the wording of a tooltip, not someone's unsupported insistence that "oh, no that tooltip that says +4 weapon damage - it's REALLY generic damage because trust me."

If Might of Cenarius' +4 weapon damage does indeed work for druids (and we know it does), then that's unquestionable precedent for +9 weapon damage enchants working as well. Weightstones? Perhaps not, since patch 2.4.0 explicitly makes weightstones work while shapeshifted, and it's unclear whether the weightstones (a) worked in Vanilla then didn't early TBC, or (b) never worked until 2.4.0.

Might of Cenarius is precedent for +9 weapon damage enchant working for druids. This is a no brainer.

**Comment by Nostalian** [comment]
absent having the actual vanilla code - not LH core code - there's no evidence that the +4 weapon damage from might of cenarius was meant to be applied to some passive weapon damage value. one could just as easily interpret +4 weapon damage as increasing the damage of whatever weapon the druid currently has equipped, a case in which clearly the druid's 'claws' wouldn't qualify as a weapon because they are not equipped by the druid per se(edited)
the burden of proof is and should be on the nay-sayers to convincingly show that might of cenarius functioned uniquely in this  way during vanilla. until that happens, and if we buy that might of cenarius worked for druids in vanilla (which most seem to do), it's sufficient precedent for +9 wpn dmg working as well

**Comment by GuybrushGit** [comment]
I'm dismissing the tooltip for a reason.
Might of Cenarius applies the SPELL_AURA_MOD_DAMAGE_DONE modifier, this is the same modifier used for example by spell damage gear or berserking buffs.
Removing the Cenarius modifier while feral makes as much sense as removing all spell damage when disarming a mage.
Vanilla comments confirm this behaviour.

**Comment by ZakPreston** [comment]
Is there any single proof that +X Weapon damage enchants didn't work in Vanilla for druids (except a couple of comments from wowhead)? This whole fuss with "fixing" weapon damage for cats and bears looks like a voluntaristic attempt to "fix a fix so that you could fix while you are fixing". 
But the real problem is that by fixing the questionable game mechanics you are literally making one of the worst specs even less playable one. I'm all for blizzlike changes, but only 100% confirmed ones, sorry.

TL;DR: by applying unconfirmed (and extremely questionable) change to one of least represented specs in vanilla you are literally making it unplayable and forcing to quit or reroll hundreds of players. It would be still understandable if the change was 100% blizzlike, but **what if you are wrong**?

**Comment by Fiskehatt** [comment]
To be honest, there is no proof that the enchants should have worked in vanilla, only evidence that points against it: several comments from various vanilla sources (not wowhead), WoWWiki article from vanilla and patch notes for 2.4.0 that proof that temporary enchants did not work before it. Just check the original issue. There also used to be even more evidence on the original nostalrius bug report but that is now sadly lost. Anyways, I cannot think of what kind of evidence would still be needed.

Moreover, it is rather weird that no one ever reported it or considered it in theorycrafting in vanilla considering that it is a rather big dps upgrade. Furthermore, it seems logical that they should not work: weapon damage has no effect in forms (or do you pick your weapon based on its damage?). 

The bug exists just because some MaNGOS developer did not originally consider ferals using weapon damage enchants. Now people falsely believe it is the blizzlike behaviour, even though the bug has been confirmed on Nostalrius, Kronos and Elysium and the evidence points against it.

**Comment by ZakPreston** [comment]
Post guides or even videos form vanilla-retail times where it's clearly visible that +X weapon damage should not work in feral forms. Otherwise you are just killing one of the least viable specs completely.

**Comment by Nostalian** [comment]
> 
only evidence that points against it: several comments from various vanilla sources (not wowhead), WoWWiki article from vanilla and patch notes for 2.4.0 that proof that **temporary enchants** did not work before it.
> 

Patch 2.4.0 is cause to believe that **temporary enchants** like weightstones and sharpening stones did not work up until that point, not +weapon damage **permanent enchants** like superior impact. Using this as an argument against permanent weapon damage enchants is really dishonest.

> 
Moreover, it is rather weird that no one ever reported that the enchants worked or considered them in theorycrafting in vanilla considering that it is a rather big dps upgrade.
>

Here's an example of a vanilla WoW feral dps theorycrafter that Taladril showed me: http://tangedyn.pbworks.com/w/page/19362776/Old%20Front%20Page
This guy is touted as one of the top feral dps theorycrafters of his time, making the OG feral dps bis list.

However, look at his helm bis list and his rotation - he doesn't even know to use wolfshead helm + furor and powershift for maximum dps!

If the top feral druid dps theorycrafters in vanilla WoW didn't realize that wolfshead helm is bis forever and powershifting with that helm + furor will always be max dps, it wouldn't surprise me if they also neglected to truly test +weapon damage enchants and instead dismissed their effectiveness solely based on how unintuitive it would be for the enchants to work.

Which brings me to your next argument:
>
 Furthermore, it seems **logical** that they should not work: weapon damage has no effect in forms (or do you pick your weapon based on its damage?).
>
Right, it seems logical. But to prove that it "makes sense" or is "logical" is not equivalent to proving that this is how Blizzard implemented it in vanilla. "This makes sense, it's logical" does not equal "this is how Blizzard implemented it in vanilla". Moreover, there's no bright line for what "makes sense" in a high-fantasy MMO-RPG; how do we decide what makes **enough** sense to be in the game and what doesn't in a world with elves, orcs, dragons, etc.? Better yet, how do we know that Blizzard even cared?

The answer is that we don't

WTB **EVIDENCE**

**Comment by Fiskehatt** [comment]
My original issue was about both temporary and permanent enchants not just permanent. Currently, all the evidence points out to: 

### 1) Temporary enchants - Did not work.

**Molten core speed guide** (http://www.ownedcore.com/forums/world-of-warcraft/world-of-warcraft-guides/4391-speed-up-molten-core.html) (10-09-2006):

> Casters can use Wizard Oil to increase their damage; melee DPS can use Sharpening Stones or Weightstones or Posions. Potions and Elixirs can also be used, but are a bit more expensive and do not last through death. Help provide them for your most effective players, but don’t insist on them unless you’re pushing hard to set a guild record… save them for content you really need the extra help on. One note for the druids out there — **only Elemental Sharpening Stones work in cat form,** and they tend to be expensive to make.

**Patch 2.4.0** (http://wowwiki.wikia.com/wiki/Patch_2.4.0):

> Weightstones and Sharpening Stones: The bonus damage from these temporary enchantments now works while shapeshifted.


### 2) Permanent enchants - Did not work.

**A stickied guide from WoW Forums** (http://web.archive.org/web/20060705020604/http://forums.worldofwarcraft.com:80/thread.aspx?fn=wow-druid&t=296661&p=1&tmp=1#post296661) (7/4/2005):

> Because we are not actually using the weapon itself to hit with while shapeshifted, 'Chance on hit' effects (eg. Crusader, Lifestealing, Fiery, Icy, Unholy) do not proc. **Impact and Striking enchants that add damage per hit also do not function.**

**Wowwiki article from November 4th 2006**  (http://wowwiki.wikia.com/wiki/Druid?direction=prev&oldid=314984):

> In Bear, Dire Bear and Cat form, the following are also true:

> Weapon DPS and **weapon +damage enchants are not used**. The weapon damage bonus from the Might of Cenarius ring does work, however.

**Discussion on WoW Allakhazam Forums** (http://wow.allakhazam.com/forum.html?forum=243&mid=115521271933777363#2) (August 2006)

> +15 Agility, +25 Agility or +15 Strength are the only options for a feral druid. 
>
>Weapon damage does not carry over, it increases the High End Damage of the weapon, which you aint using, only its stats. 
>
>So, for Recap 
>
>Fiery Weapon, Icy Chill, Unholy Weapon, Crusader, 1h Striking, 2h striking, Demonslaying etc.. Do not work in Shapeshifted forms. 

>Might of Cenarius works in forms. It adds to weapon damage, which happens to be your claw. **+4 dmg enchant on your mace doesn't work, because you're not using your mace.**

http://wow.allakhazam.com/forum.html?forum=21&mid=113950310940335407 (Feb 2006)

> Don't get +damage either on a weapon... goes without saying I hope, but that won't do anything for you if you're not hitting with that weapon, even though it's not a true proc.

**Thottbot comments on +9 weapon damage enchants** (https://web.archive.org/web/20070515223857/http://www.thottbot.com:80/s20030) (September 2006):

> so is this helpful to feral druids playing as a cat, when their atk speed is set at 1 sec regardless of the weapons speed?

> Sadly not. **This enchant has no effect on druids in feral forms**. The damage range of your equipped weapon isn't carried over to your claws. 
>
> The good news is that the +15str enchant will benefit you, as will any other enchant that improves your stats.
 
**Thottbot comments on +5 weapon damage enchant** (https://web.archive.org/web/20070319182715/http://www.thottbot.com:80/s20031) (April 2006):

> no, druid animal forms does not use weapon damage range, damage enchants or procs.

There are also several guides that list agility and strength enchants as the only options for feral druids.

### 3) Might of Cenarius - Worked

See the earlier comments, especially the one from WoWWiki.

---

To sum it up, there is **a lot of evidence pointing out that weapon enchants did not work, none that they worked**. Wrath of Cenarius works only because of the way it is implemented as explained by GuybrushGit, it does not prove that the enchants worked.

---

Regarding Tangedyn, he must have known about powershifting: his [Druid DPS calculator](http://tangedyn.pbworks.com/w/page/19362767/Cat%20Simulator) calculates both non-shifting and power-shifting DPSs and uptime:

http://i.pi.gy/rjpbm.png

He also knew about the ring working as it can be seen in his BIS gearlist:

http://tangedyn.awardspace.com/cat-dps-items.htm

Even when knowing so much about feral dps, he did not use a weapon damage enchant but +15 str on his main hand:

http://web.archive.org/web/20060420033911/http://www.ctprofiles.net:80/12764

Anyways, it is possible that the whole druid community missed that weapon enchants worked, however, nothing currently points to that. Feel free to find **evidence**.

**Comment by Nostalian** [comment]
>Regarding Tangedyn, he must have known about powershifting: his Druid DPS calculator calculates both non-shifting and power-shifting DPSs and uptime: http://i.pi.gy/rjpbm.png

Not sure where in the image you're getting this. Yes, there are two different dps values in the image you provided - DPS-Normal and DPS-Shifting -  but if he was truly simulating powershifting with a wolfshead helm I'd expect DPS-Shifting to be significantly above DPS-Normal, which it's not. Furthermore, in the description of his cat sim here: 

http://tangedyn.pbworks.com/w/page/19362767/Cat%20Simulator 

there is no indication that it simulates cat dps both with and without powershifting. The closest thing mentioned is this: 

>You can set the length of each run to 0, this will terminate the run right after the first finisher and compute the DPS from there. This can be used for extreme burst damage computation.

I browsed the rest of his website, even searching for the term "shift", and couldn't find anything indicating that he actually knew to powershift with a wolfshead helm; rather, the image of his simulator suggested that whatever notion he had of a dps rotation that involved any "shifting" was clearly something different enough to not constitute a dps increase over a non-"shifting" rotation. The Average Combo Time value of 32.4 seconds reinforces this, as if he was truly simulating powershifting with a wolfshead helm there'd be some smaller Average Combo Time value in the sim that he uses, and I don't see one in the image you provided. Finally, there's still no wolfshead helm on his cat dps bis list (nor is there one in this cat dps guide that your third quote is from: http://web.archive.org/web/20060705020604/http://forums.worldofwarcraft.com:80/thread.aspx?fn=wow-druid&t=296661&p=1&tmp=1#post296661  - doesn't make me very confident in the rest of the "guides" you claim suffice as evidence!)

So I'll reiterate my point:

>If the top feral druid dps theorycrafters in vanilla WoW didn't realize that wolfshead helm is bis forever and powershifting with that helm + furor will always be max dps, it wouldn't surprise me if they also neglected to truly test +weapon damage enchants and instead dismissed their effectiveness solely based on how unintuitive it would be for the enchants to work.

Now observe the rhetoric used by some of your sources when they claim that +weapon damage enchants don't work for ferals (pay attention to the bolded text):

>Don't get +damage either on a weapon... **goes without saying I hope**, but that won't do anything for you **if you're not hitting with that weapon**, even though it's not a true proc.

>Sadly not. This enchant has no effect on druids in feral forms. **The damage range of your equipped weapon isn't carried over to your claws.**

>Weapon damage does not carry over, it increases the High End Damage of the weapon, **which you aint using**, only its stats.

Coupled with my first point, some of your quotes only further suggest that +weapon damage enchants were not rigorously tested in Vanilla WoW and that contemporary players merely appealed to intuition/ logic when considering whether or not the enchants would work. To them, it was something that went "without saying" and logically followed from the fact that cat dps don't use they're equipped weapons while in form. 

So **no**, there is not 
>a lot of evidence pointing out that weapon damage enchants did not work, none that they worked.

There _is_ a lot of people who I've already shown were behind on cat dps theorycrafting that **assert** that the +weapon damage enchants don't work. They **don't** cite any testing methodologies, they **don't** claim to have looked at the character sheet with and without the enchant, and they certainly **didn't** have access to the source code to confirm that the weapon damage from might of cenarius and the weapon damage from +weapon damage enchants were added to different values (understandable obviously - lol - but nevertheless not helpful to their argument).

So I still don't think that sufficient evidence exists to implement this change.

>Btw, if you are worried about feral dps, you can go look at Wortgewandt's video. He does pretty good dps in raids with just an agility enchant and crit stones. Claiming that feral druids become completely useless with the fix is an exaggeration.

Yeah I definitely don't believe that ferals become "useless" without being able to use +weapon damage enchants or weightstones. As someone who raids as feral dps, I'm confident I'd still be able to push good numbers. I'm mostly continuing this discussion because I think it's an important discussion to be had, especially since I genuinely don't think that enough evidence for the change you support exists and it would be really lame for a change supported by nothing more than shoddy anecdotes to be implemented.

**Comment by Lorilay2** [comment]
Just chiming in here as someone that specifically tested this in the "Cat DPS Megathread" post on Elitist Jerks. No, weightstones or +damage enchants did not work in vanilla. Multiple druids (including me) tested it on white quality staves and saw no damage change with p<0.01. It's for that reason that the optimal dps enchants were listed as +25 agility and +15 strength, rather than +5 or +9 damage. If they did work, their effectiveness would have been obvious to even the most basic of math comprehension.

**Comment by AliUDmov** [comment]
@Fiskehatt thank you for this compilation.

There are even more sources on http://thedruidsgrove.org/archive/wow/ that suggest that +weapondamage enchants and weightstones did not work until TBC.

There is not a single source 2004-2006 that suggests it had worked. There are around 20-30 sources that state it had not worked. @ZakPreston it's not about killing a spec, it's just about being true to Vanilla, without custom server shenanigans. Downvoting @GuybrushGit is quite unfair, we should be grateful that he looks into this issue and tries to fix it. :)

**Comment by Fiskehatt** [comment]
Just adding this old WoW blue tracker "bug" post to the list (http://blue.cardplace.com/cache/wow-customer-service/38614.htm):

> Poster: Dubh at 6/3/2006 12:30:58 PM PDTSubject: + weapon damage doesnt work for feral druids

>  i have a +5 dmg to my thornstone. although the dps while in cat and bear form doesn't increase at all. dense weightstones have the same effect. zero. i thought that it was just the chance on hit/prod enchants that didnt work. should these increase feral dps? they don't for me.

This is answered by a blizz employee:

> Poster: Kaone at 6/3/2006 12:39:09 PM PDTSubject: Re: + weapon damage doesnt work for feral druids

> Hello Dubh, 
>
> In druid feral forums the DPS of the equipped weapon is not the determining factor when calculating your characters DPS. This is entirely based on the stats on your character at the time. This means while stat bonuses on your weapon or on your enchant will increase your DPS, any +X damage enchants will not.

File: src/game/StatSystem.cpp [hunk 1]
```diff
@@ -414,12 +415,14 @@ void Player::CalculateMinMaxDamage(WeaponAttackType attType, bool normalized, fl
...
             weapon_mindamage = lvl * 0.85f * att_speed;
             weapon_maxdamage = lvl * 1.25f * att_speed;
+            total_value = 0.0f;                             // remove benefit from weapon enchants
         }
     }
     else if (!CanUseEquippedWeapon(attType))                // check if player not in form but still can't use weapon (broken/etc)
     {
         weapon_mindamage = BASE_MINDAMAGE;
         weapon_maxdamage = BASE_MAXDAMAGE;
+        total_value = 0.0f;
```
Comment (thepparker) [inline src/game/StatSystem.cpp]: I'm not a huge fan of this. If we could split weapon-specific enchants out of the weapon damage and then re-apply them to the total in this method it would probably be better.

Comment (GuybrushGit) [reply src/game/StatSystem.cpp]: Something like this? https://github.com/GuybrushGit/server-1/commit/6d7dae2f69035ef192f0c2322b4a27ff0894aa1b 

That's one way to do it I guess.

Comment (thepparker) [reply src/game/StatSystem.cpp]: I was thinking more along the lines of storing them in separate attributes for the weapon, but thinking about it, it might be a bit overkill. I suppose a generic method that does what you've just written would do the job.

Comment (GuybrushGit) [reply src/game/StatSystem.cpp]: What I've done in this pull request is just that. I'm storing the generic damage in UNIT_MOD_DAMAGE_PHYSICAL so all that's left in UNIT_MOD_DAMAGE_MAINHAND / OFFHAND is the enchant values.

Comment (thepparker) [reply src/game/StatSystem.cpp]: Ah, right you are. For some reason I was thinking weapon damage was stored in UNIT_MOD

**Comment by Kittnz** [comment]
https://github.com/LightsHope/server/issues/1031
